crates/ol/mempool/src/state.rs

```diff
@@ -0,0 +1,1589 @@
+    /// StateAccessor is created when tip changes in `set_current_tip()`.
+    state_accessor: Option<Arc<OLState>>,
+
+    /// Mempool statistics.
+    stats: OLMempoolStats,
+}
+
+impl MempoolServiceState {
+    /// Create new mempool service state.
+    #[expect(dead_code, reason = "will be used via builder")]
+    pub(crate) fn new(
+        config: OLMempoolConfig,
+        storage: Arc<NodeStorage>,
+        current_tip: OLBlockCommitment,
+    ) -> Self {
+        let ctx = Arc::new(MempoolContext::new(config, storage));
+        Self::new_with_context(ctx, current_tip)
+    }
+
+    /// Create new mempool service state with an existing context.
+    /// Used for testing.
+    pub(crate) fn new_with_context(
+        ctx: Arc<MempoolContext>,
+        current_tip: OLBlockCommitment,
+    ) -> Self {
+        Self {
+            ctx,
+            entries: HashMap::new(),
+            ordering_index: BTreeMap::new(),
+            by_account: HashMap::new(),
+            next_insertion_id: 0,
+            current_tip,
+            pending_seq_no: HashMap::new(),
+            state_accessor: None,
+            stats: OLMempoolStats::default(),
+        }
+    }
+
+    /// Load existing transactions from database.
+    ///
+    /// Deserializes and validates each transaction by calling `add_transaction`.
+    /// Invalid transactions are skipped and removed from the database.
+    pub(crate) async fn load_from_db(&mut self) -> OLMempoolResult<()> {
+        let all_txs = self.ctx.storage.mempool().get_all_txs()?;
+
+        let mut loaded_count = 0;
+        let mut skipped_count = 0;
+
+        for tx_data in all_txs {
+            // Parse transaction from bytes
+            let tx = match ssz::Decode::from_ssz_bytes(&tx_data.tx_bytes) {
+                Ok(tx) => tx,
+                Err(e) => {
+                    // Skip malformed transaction and remove from DB
+                    tracing::warn!(
+                        txid = ?tx_data.txid,
+                        error = ?e,
+                        "Skipping malformed transaction from database"
+                    );
+                    let _ = self.ctx.storage.mempool().del_tx(tx_data.txid);
+                    skipped_count += 1;
+                    continue;
+                }
+            };
+
+            // Validate and add transaction
+            match self.add_transaction(tx).await {
+                Ok(_) => {
+                    loaded_count += 1;
+                }
+                Err(e) => {
+                    // Skip invalid transaction and remove from DB
+                    tracing::warn!(
+                        txid = ?tx_data.txid,
+                        error = ?e,
+                        "Skipping invalid transaction from database"
+                    );
+                    let _ = self.ctx.storage.mempool().del_tx(tx_data.txid);
+                    skipped_count += 1;
+                }
+            }
```

Comment:
because we don't allow gaps in seq no, we should validate tx in a particular order: if we have seq no `1,2` then the order of validation would result in different results.

I advise again to rethink should we not allow gaps in the mempool design.